### PR TITLE
enhancement/peer-id-bytes

### LINKED ISSUE #40: Change PeerID from ECSDA to Bytes
We need to turn it to bytes to deepequal anyway, might as well just use bytes native.

### PR DESCRIPTION
**Pull request by decanus**
closes #40 

@oskarth @adambabik we need to discuss what size the array should have. Right now 64 was taken, this is completely arbitrary though.

### CODE REVIEW

File: peerid.go [hunk 1]
```diff
@@ -7,12 +7,14 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
-type PeerId ecdsa.PublicKey
+type PeerId [64]byte
 
-func (p PeerId) ToBytes() []byte {
-	if p.X == nil || p.Y == nil {
+func PublicKeyToPeerID(k ecdsa.PublicKey) *PeerId {
```
Comment (adambabik) [inline peerid.go]: I don't think the pointer here is necessary. 
```suggestion
func PublicKeyToPeerID(k ecdsa.PublicKey) PeerId {
```

File: peerid.go [hunk 2]
```diff
@@ -7,12 +7,14 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
-type PeerId ecdsa.PublicKey
+type PeerId [64]byte
```
Comment (adambabik) [inline peerid.go]: Usually in Go, the naming convention for shortcuts and keywords is to uppercase them:
```suggestion
type PeerID [64]byte
```

similarly with `HTTPStatus` or `MarshalJSON` etc.

File: peerid.go [hunk 3]
```diff
@@ -7,12 +7,14 @@ import (
...
 
-func (p PeerId) ToBytes() []byte {
-	if p.X == nil || p.Y == nil {
+func PublicKeyToPeerID(k ecdsa.PublicKey) *PeerId {
+	if k.X == nil || k.Y == nil {
 		return nil
 	}
 
-	return elliptic.Marshal(crypto.S256(), p.X, p.Y)
+	p := &PeerId{}
```
Comment (adambabik) [inline peerid.go]:
```suggestion
	var p PeerID
```

File: simulation/simulation.go [hunk 4]
```diff
@@ -153,7 +153,7 @@ func Calc(count uint64, epoch int64) int64 {
 
 func peerId() mvds.PeerId {
 	key, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-	return mvds.PeerId(key.PublicKey)
+	return *mvds.PublicKeyToPeerID(key.PublicKey)
```
Comment (adambabik) [inline simulation/simulation.go]: And you can avoid dereferencing a pointer here if you don't return a pointer from `PublicKeyToPeerID`.

**Review by adambabik — APPROVED** [review]

**Comment by decanus** [comment]
@adambabik do you think the length is ok?

File: peerid.go [hunk 5]
```diff
@@ -7,12 +7,15 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
-type PeerId ecdsa.PublicKey
+type PeerID [64]byte
 
-func (p PeerId) ToBytes() []byte {
-	if p.X == nil || p.Y == nil {
-		return nil
+func PublicKeyToPeerID(k ecdsa.PublicKey) PeerID {
```
Comment (adambabik) [inline peerid.go:11]: Actually, there is https://github.com/ethereum/go-ethereum/blob/master/crypto/crypto.go#L142-L147 which does exactly what you need. But it returns a slice not an array.

**Comment by adambabik** [comment]
@decanus the length of what?

**Comment by decanus** [comment]
@adambabik the array, 64 byte.

**Review by oskarth — COMMENTED** [review]
Outside of public key (32?), what do we imagine putting in here?

cc @cammellos who raised this concern

**Comment by cammellos** [comment]
I guess it depends on the use cases, but widely accepted schemes are uuid for example, which is 16 bytes, and would fit fine in the 64 (for sure 64 is plenty).

In our case we could use the compressed version, which is 33 bytes, if we want to reduce the size. But more generally unless we need to recover the key, we can just use an hash of the key (16, 32, 64 bytes, up to us)

**Comment by decanus** [comment]
@cammellos I'll merge it with 64, we can change it later.

**Pull request merged**